core/templates/components/ck-editor-helpers/ck-editor-copy-content-service.ts

```diff
@@ -0,0 +1,165 @@
+  private copyEventEmitter = new EventEmitter<CkEditorCopyEvent>();
+  private ckeditorToSubscription: {[id: string]: Subscription} = {};
+
+  copyModeActive = false;
+
+  constructor(private htmlEscaperService: HtmlEscaperService) {}
+
+  /**
+   * Traverses up and down element ancestors/descendants, searching for widget
+   * tags.
+   * @param {HTMLElement} target The element target that contains or is a
+   *  descendant of a widget, or a plain HTML element/group.
+   * @returns {CkEditorCopyEvent} Returns an object of type CkEditorCopyEvent,
+   *  where rootElement is the root ancestor of target and is
+   *  always a child of angular-html-bind. containedWidgetTagName is the tag
+   *  name of the widget if found in ancestors or descendants.
+   */
+  private _handleCopy(target: HTMLElement): CkEditorCopyEvent {
+    let containedWidgetTagName;
+    let currentElement = target;
+
+    while (
+      currentElement.parentElement.tagName !== 'ANGULAR-HTML-BIND'
+    ) {
+      const currentTagName = currentElement.tagName.toLowerCase();
+      if (currentTagName.includes('-noninteractive-')) {
+        containedWidgetTagName = currentTagName;
+      }
+      currentElement = currentElement.parentElement;
+    }
+
+    let descendents = Array.from(target.childNodes);
+    while (descendents.length !== 0) {
+      let currentDescendent = descendents.shift();
+      const currentTagName = currentDescendent.nodeName.toLowerCase();
+      if (currentTagName.includes('-noninteractive-')) {
+        containedWidgetTagName = currentTagName;
+        break;
+      }
+
+      descendents = [
+        ...descendents,
+        ...Array.from(currentDescendent.childNodes)
+      ];
+    }
+
+    return { rootElement: currentElement, containedWidgetTagName };
+  }
+
+  /**
+   * Detects if element is to be copied as a widget or as a HTML element, and
+   * inserts into editor.
+   * @param {CKEDITOR.editor} editor The editor in which to create widget or
+   *  insert HTML.
+   * @param {HTMLElement} element The element to be copied.
+   * @param {string|undefined} containedWidgetTagName The name of the widget
+   *    in which element contains, if present.
+   */
+  private _handlePaste(
+      editor: CKEDITOR.editor | Partial<CKEDITOR.editor>,
+      element: HTMLElement,
+      containedWidgetTagName?: string
+  ) {
+    let tagName = (
+      containedWidgetTagName || element.tagName.toLowerCase());
+    let html = element.outerHTML;
+
+    if (
+      !containedWidgetTagName && !tagName.includes('-noninteractive-')
+    ) {
+      editor.insertHtml(html);
+    } else {
+      const widgetName = tagName.replace('-noninteractive-', '');
+      const valueMatcher = /(\w+)(-with-value=")([^"]+)(")/g;
+      let match;
+      let startupData: {[id: string]: string} = {};
+
+      while ((match = valueMatcher.exec(html)) !== null) {
+        const key = match[1];
+        const value = match[3].replace(/&amp;/g, '&');
```

Comment:
```escapedStrToUnescapedStr``` matches for &:
```
escapedStrToUnescapedStr(value: string): string {
    return String(value)
      .replace(/&quot;/g, '"')
      .replace(/&#39;/g, '\'')
      .replace(/&lt;/g, '<')
      .replace(/&gt;/g, '>')
      .replace(/&amp;/g, '&');
  }
```
The problem is that & is represented as &amp; in the html. 

So if we call ```escapedStrToUnescapedStr('&amp;quot;&amp;quot;')```, the return value will be '&quot;&quot;', instead of "" like we want.

So I just preprocess the &amp; so that ```escapedStrToUnescapedStr``` matches correctly.

Reply:
I'm still not getting this (sorry!), I mean why do we want to process the data here I was expecting the interaction's to handle this. Can you give me a real example with any interaction's attribute (that will help!).

Reply:
Can you add a comment explaining the reason for the replacement?

Reply:
Sure

Reply:
@DubeySandeep we have to extract the widget data from the html. For some html:
```
<oppia-noninteractive-image alt-with-value="&amp;quot;&amp;quot;" capt' +
      'ion-with-value="&amp;quot;Banana&amp;quot;" filepath-with-value="&amp;' +
      'quot;img_20200630_114637_c2ek92uvb8_height_326_width_490.png&amp;quot;' +
      '"></oppia-noninteractive-image>
```

We are trying to capture all of the widget data, which are the attributes suffixed with `-with-value`. Ckeditor expects is to get a dictionary of the startup data. So in this case we would want 
```
{
    alt: "",
    caption: "Banana",
    filepath: "img_20200630_114637_c2ek92uvb8_height_326_width_490.png"
}
```
Does that clear things up?